rye/src/cli/publish.rs

```diff
@@ -0,0 +1,235 @@
+    repository_url: Url,
+    /// An access token used for the upload.
+    #[arg(long)]
+    token: Option<String>,
+    /// Sign files to upload using GPG.
+    #[arg(long)]
+    sign: bool,
+    /// GPG identity used to sign files.
+    #[arg(short, long)]
+    identity: Option<String>,
+    /// Path to alternate CA bundle.
+    #[arg(long)]
+    cert: Option<PathBuf>,
+    /// Enables verbose diagnostics.
+    #[arg(short, long)]
+    verbose: bool,
+    /// Turns off all output.
+    #[arg(short, long, conflicts_with = "verbose")]
+    quiet: bool,
+}
+
+pub fn execute(cmd: Args) -> Result<(), Error> {
+    let output = CommandOutput::from_quiet_and_verbose(cmd.quiet, cmd.verbose);
+    let venv = ensure_self_venv(output)?;
+    let project = PyProject::discover()?;
+
+    // Get the files to publish.
+    let files = match cmd.dist {
+        Some(paths) => paths,
+        None => vec![project.workspace_path().join("dist").join("*")],
+    };
+
+    // a. Get token from arguments and offer encryption, then store in credentials file.
+    // b. Get token from ~/.rye/credentials keyed by provided repository and provide decryption option.
+    // c. Otherwise prompt for token and provide encryption option, storing the result in credentials.
+    let repository = &cmd.repository;
+
+    // If -r is pypi but the url isn't pypi then bail
+    if repository == "pypi" && cmd.repository_url.domain() != Some("pypi.org") {
+        bail!("invalid pypi url {} (use -h for help)", cmd.repository_url);
+    }
+
+    let mut credentials = get_credentials()?;
+    credentials
+        .entry(repository)
+        .or_insert(Item::Table(Table::new()));
+
+    let token = if let Some(token) = cmd.token {
+        let secret = Secret::new(token);
+        let maybe_encrypted = prompt_maybe_encrypt(&secret)?;
+        let maybe_encoded = maybe_encode(&secret, &maybe_encrypted);
+        credentials[repository]["token"] = Item::Value(maybe_encoded.expose_secret().into());
+        write_credentials(&credentials)?;
+
+        secret
+    } else if let Some(token) = credentials
+        .get(repository)
+        .and_then(|table| table.get("token"))
+        .map(|token| token.to_string())
+        .map(escape_string)
+    {
+        let secret = Secret::new(token);
+
+        prompt_maybe_decrypt(&secret)?
+    } else {
+        eprintln!("No access token found, generate one at: https://pypi.org/manage/account/token/");
+        let token = prompt_for_token()?;
+        if token.is_empty() {
+            bail!("an access token is required")
+        }
+        let secret = Secret::new(token);
+        let maybe_encrypted = prompt_maybe_encrypt(&secret)?;
+        let maybe_encoded = maybe_encode(&secret, &maybe_encrypted);
+        credentials[repository]["token"] = Item::Value(maybe_encoded.expose_secret().into());
+        write_credentials(&credentials)?;
+
+        secret
+    };
+
+    let mut publish_cmd = Command::new(venv.join("bin/python"));
```

Comment:
This should use the new utility function that makes this work on windows (`get_venv_python_bin`).

Reply:
Added in f6dc0795f55ad2e952853249b9f9cf0479ae99b8

Reply:
Oof I can fix the commit message if you'd like